ls1intum/Artemis
# [Enhancement] Improvements on Reactions

### Checklist

### Motivation and Context


I had to fix a bug that was found during testing, where a @MerlinMehmed  could add a reaction twice by selecting the emoji with some double clicks - which I could not reproduce.
As I checked the related code, I did some minor improvements preventing this unintended behavior and changed the current implementation w.r.t how/if the creation date of a post/answer post/reaction was built and persisted.
Added afterwards: During testing a saw that for the page discussion section, when loading took a while, I saw "Loading" and "No posts found" at the same time. I had to adapt the html file to show "No posts found" only after loading was finished. I forgot that in a previous PR.

### Description

Client Improvements:
- Reaction selector overlay is closed whenever the users clicks somewhere outside (previously one had to click on the emoji button that opened the selector)
- The emoji button that opened the selector has a different styling (active) when the selector overlay is open
- The client does not provide creation dates for posts and answer posts as these are added on the server-side

Server Improvements:
- Reactions creations date where not created/persisted, now they are
- Creation dates for posts and answer posts are created at server-side
- We now have server-side checks that a reaction can not be duplicated, i.e. a user cannot post the same reaction twice, most importantly, cannot vote twice

### Steps for Testing


1. Log in to Artemis as any user
2. Navigate to a course with posts (testserver 1 has a course with many posts, its id is 477, you could use e.g. test user up to id 15)
3. create a post, no errors should occur, created date should be set correctly
4. create an answer post, no errors should occur, created date should be set correctly
5. edit a post, no errors should occur, creation date should be the same
6. edit an answer post, no errors should occur, creation date should be the same
7. react on a post and answer post, click on the reaction again -> should be removed; also very quickly double clicking/selecting a reaction should not lead to creating a reaction twice as a use

### Review Progress





- Code Review
- Manual Tests

### Test Coverage



One test for checking that duplicated reactions are prevented was added, existing tests are were adapted slightly
ReactionResource and ReactionService have 100% instruction coverage.
